### PR TITLE
Add statsdreceiver Unixgram Support

### PR DESCRIPTION
**Pull request by michaelli321**
#### Description

Adds `unixgram` transport for the `statsdreceiver`. Additionally, creates a new `packetServer` base class for both `UDS` and `UDP*` transport types 

#### Link to tracking issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21385

#### Testing

Added a unit test

### CODE REVIEW

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (dmitryax) [inline receiver/statsdreceiver/internal/transport/uds_server.go:26]: Why do we have to remove the socket here? And shouldn't we handle the error here?

Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/uds_server.go:26]: We need to remove the socket in case the socket file wasn't properly cleaned up

> And shouldn't we handle the error here?

Yes good catch, I'll add that now

**Comment by michaelli321** [comment]
@ChrsMark @jmacd @dmitryax Do you have an estimate on when someone could get around to a review?

File: receiver/statsdreceiver/internal/transport/client/client.go
Comment (ChrsMark) [inline receiver/statsdreceiver/internal/transport/client/client.go:54]: I see we don't document the `transport` setting at all but I wonder if that's a good chance to do it since we are on it.

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (ChrsMark) [inline receiver/statsdreceiver/internal/transport/uds_server.go:50]: Wouldn't be safer to just fail here instead of "force"-removing the socket? What if this socket is already in use by another Collector instance or just another application?

File: receiver/statsdreceiver/internal/transport/udp_server.go
Comment (ChrsMark) [inline receiver/statsdreceiver/internal/transport/udp_server.go:40]: This refactoring makes sense but maybe would be better if it's done in a standalone/different PR instead of mixing it with this addition: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#refactoring-work

File: receiver/statsdreceiver/receiver.go
Comment (ChrsMark) [inline receiver/statsdreceiver/receiver.go:53]: Will this default be documented in the receiver's docs for the users?

**Review by ChrsMark — COMMENTED** [review]
Overall the enhancement looks good but I left some comments with things we should consider/tackle before merge.

@dmitryax @jmacd PTAL

File: receiver/statsdreceiver/internal/transport/udp_server.go
Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/udp_server.go:40]: Hmm I see. I would argue that this refactoring work is necessary for this functionality. Otherwise, we'd be unnecessarily duplicating code for the new functionality.

If you feel strongly about this, I can make the change.

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/uds_server.go:50]: Hmm that's fair. At my previous company, we deployed the otel collector as a daemonset so we used this pattern in case the socket wasn't properly cleaned up from an old instance of the collector. What are your thoughts in that case?

File: receiver/statsdreceiver/internal/transport/client/client.go
Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/client/client.go:54]: The `transport` setting comes from the `go.opentelemetry.io/collector/config/confignet` package where it's documented [here](https://pkg.go.dev/go.opentelemetry.io/collector/config/confignet#section-readme)

Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/client/client.go:54]: In light of [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36608#discussion_r1880020678), I've added some specific documentation for the `statsdreceiver`

File: receiver/statsdreceiver/receiver.go
Comment (michaelli321) [reply receiver/statsdreceiver/receiver.go:53]: Thanks, added some documentation to the README

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (ChrsMark) [reply receiver/statsdreceiver/internal/transport/uds_server.go:50]: I don't think it should be responsibility of the new collector instance itself to clean this up. If somebody would want to automate this, they could easily hook an init container or any kind of pre-run bash script to take care of it on administrative level.

Having this being part of the Collector would even allow to someone using it maliciously.

File: receiver/statsdreceiver/internal/transport/udp_server.go
Comment (ChrsMark) [reply receiver/statsdreceiver/internal/transport/udp_server.go:40]: Yeap, I see. I'm fine keeping it as is.

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (dmitryax) [reply receiver/statsdreceiver/internal/transport/uds_server.go:50]: I agree with @ChrsMark here. An init container can take care of that. I don't think collector should assume write access to the filesystem

Comment (dmitryax) [reply receiver/statsdreceiver/internal/transport/uds_server.go:50]: Can this be a configurable optional behavior?

**Review by dmitryax — COMMENTED** [review]
The code looks good to me. Waiting for a conclusion the socket file discussion

File: receiver/statsdreceiver/internal/transport/uds_server.go
Comment (michaelli321) [reply receiver/statsdreceiver/internal/transport/uds_server.go]: Makes sense. I've removed this check

File: receiver/statsdreceiver/README.md
Comment (ChrsMark) [inline receiver/statsdreceiver/README.md:22]: `endpoint` used to be under the required settings, right?

Has this changed? Maybe we can keep it there to avoid confusion with existing users unless it's wrong.

**Review by ChrsMark — APPROVED** [review]
LGTM

I left one minor clarification question for the docs.

**Comment by ChrsMark** [comment]
> The code looks good to me. Waiting for a conclusion the socket file discussion

@dmitryax the socket deletion was removed, please take a look.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
